packages/client/src/features/tools/edge-creation/edge-creation-tool.ts

```diff
@@ -128,33 +141,74 @@ export class EdgeCreationToolMouseListener extends DragAwareMouseListener {
         return this.target !== undefined;
     }
 
-    override mouseOver(target: SModelElement, event: MouseEvent): Action[] {
+    override mouseOver(target: SModelElement, event: MouseEvent): (Action | Promise<Action>)[] {
         const newCurrentTarget = findParentByFeature(target, isConnectable);
         if (newCurrentTarget !== this.currentTarget) {
             this.currentTarget = newCurrentTarget;
             if (this.currentTarget) {
                 if (!this.isSourceSelected()) {
                     this.allowedTarget = this.isAllowedSource(newCurrentTarget);
                 } else if (!this.isTargetSelected()) {
-                    this.allowedTarget = this.isAllowedTarget(newCurrentTarget);
-                }
-                if (this.allowedTarget) {
-                    const action = !this.isSourceSelected()
-                        ? cursorFeedbackAction(CursorCSS.EDGE_CREATION_SOURCE)
-                        : cursorFeedbackAction(CursorCSS.EDGE_CREATION_TARGET);
-                    return [action];
+                    // Temporarily mark the target as invalid while we check for a proper result,
+                    // so a fast-clicking user doesn't get a chance to create the edge in the meantime.
+                    this.allowedTarget = false;
+                    const actions = this.isAllowedTarget(newCurrentTarget).then(allowedCheck => {
+                        // Make sure we didn't change the target element while
+                        // checking for valid target.
+                        if (allowedCheck.targetElement === this.currentTarget) {
+                            this.allowedTarget = allowedCheck.allowed;
+                            return this.updateEdgeFeedback();
+                        }
+                        // FIXME Is there a proper no-op action? We can't return <undefined> or [] here
```

Comment:
I wondering whether we really need to add this to the return result of the `mouseOver` method. We have access to the action dispatcher, so couldn't we simply dispatch the action returned from `this.updateEdgeFeedback()` directly with the dispatcher?

Reply:
That would be an option indeed; I tried to stick to the defined API to return the feedback actions. But since the mouse tool only delegates to the ActionDispatcher, I guess it'd be fine to do it directly in the listener.